ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,264 @@
+import os
+import re
+import sys
+from typing import TYPE_CHECKING
+
+import graphql
+from graphql import MiddlewareManager
+from graphql.error import GraphQLError
+from graphql.execution import ExecutionResult
+from graphql.language.source import Source
+
+from ddtrace import config
+from ddtrace.constants import ANALYTICS_SAMPLE_RATE_KEY
+from ddtrace.constants import SPAN_MEASURED_KEY
+from ddtrace.internal.compat import stringify
+from ddtrace.internal.utils import ArgumentError
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.utils import set_argument_value
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.internal.utils.version import parse_version
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+from ddtrace.pin import Pin
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+if TYPE_CHECKING:
+    from typing import Callable
+    from typing import Dict
+    from typing import Iterable
+    from typing import List
+    from typing import Tuple
+    from typing import Union
+
+    from ddtrace import Span
+
+
+_graphql_version_str = getattr(graphql, "__version__")
+graphql_version = parse_version(_graphql_version_str)
+
+if graphql_version < (3, 0):
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+config._add("graphql", dict(_default_service="graphql"))
+
+
+def _graphql_resolvers_enabled():
+    """
+    Patching graphql resolvers may produce large traces.
+    DD_TRACE_GRAPHQL_RESOLVERS_ENABLED is disabled by default.
+    """
+    return asbool(os.getenv("DD_TRACE_GRAPHQL_RESOLVERS_ENABLED", default=False))
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+    setattr(graphql, "_datadog_patch", True)
+    Pin().onto(graphql)
+
+    for module_str, func_str, wrapper in _get_patching_candidates():
+        module = sys.modules[module_str]
```

Comment:
The `import graphql` statement at the top of the file ensure that these modules can be found in sys modules (all patched functions are imported in `graphql.__init__.py` either directly or indirectly but this could change in a future release). That's why the current approach works.

The main issue with using `__import__()` is that [graphql.graphql](https://github.com/graphql-python/graphql-core-legacy/blob/v2.0.0/graphql/graphql.py) is a module and [graphq.graphql()](https://github.com/graphql-python/graphql-core-legacy/blob/v2.0.0/graphql/__init__.py#L28) is a function in `graphql.__init__.py`. 

Because of this: 
`__import__("graphql.__init__") == __import__("graphql") == __import__("graphql.graphql")` 

To avoid this issue this issue I use sys.modules to get the reference to the `graphql.graphql` module. 

I run into a similar issue with the `graphql.execution.execute` module and the `graphql.execution.execute()` function. This whole thing is a mess 😭 .